### PR TITLE
Add Thunder as a Dynamo backend

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds `thunder.dynamo.ThunderCompiler` meant to be used as a `backend=` option to `torch.compile`/`torch._dynamo.optimize`. Only the most basic functionality is included in this PR. Thunder may not support the graphs generated by Dynamo, in this case, an (possibly scary) error will be thrown. 

Our main end-to-end benchmark script is modified to allow invoking Dynamo preprocessing with a `dynamo+` string passed to the `--compile` option. With Dynamo preprocessing PyTorch's FSDP and DDP are used for distributed training.

See https://github.com/Lightning-AI/lightning-thunder/issues/915 and the link to the design doc in the issue.

### CODE REVIEW

File: thunder/benchmarks/benchmark_litgpt.py [hunk 1]
```diff
@@ -393,7 +395,15 @@ def setup_compile(self, model):
 
                 executors.insert(0, transformer_engine_ex)
 
-            model = thunder.jit(model, executors=executors)
+            if "dynamo" in self.compile:
+                backend = ThunderCompiler(executors=executors)
+                # Because Lightning Fabric is imported in this script it monkey patches the torch.compile function
```
Comment (mruberry) [inline thunder/benchmarks/benchmark_litgpt.py:400]: Nice comment

File: thunder/dynamo/compiler.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import torch
+
+from collections import OrderedDict
+
+
+class ThunderCompiler:
+    def __init__(self, **thunder_options):
+        """
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:8]: Nice doc string

File: thunder/dynamo/compiler.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            ...     x = torch.sin(x)
+            ...     if x.sum() > 0:
+            ...         return x + 1
+            ...     else:
+            ...         return x - 1
+            >>> out = func(x)
+        """
+        from thunder import ThunderModule
+
+        # Thunder-compiled functions should be readily available for inspection
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:32]: Nice comment

File: thunder/dynamo/compiler.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        # functions in the list will be the same as the order in which they were
+        # compiled. In addition, we will store a mapping from the ThunderModule
+        # to the GraphModule that was passed to ThunderCompiler. This will allow
+        # us to inspect the GraphModule that was compiled by Thunder.
+        self.thunder_fns: list[ThunderModule] = []
+        self.thunder_to_gm: OrderedDict[ThunderModule, torch.fx.GraphModule] = OrderedDict()
+
+        self.thunder_options = thunder_options
+
+        # There will be pieces of Dynamo IR that Thunder cannot compile, so we
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:43]: Nit: maybe make it clearer this is future work / TODO?

File: thunder/dynamo/compiler.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        # 3. If the whole GraphModule is not supported, record the reasons why.
+
+    def __call__(self, gm: torch.fx.GraphModule, sample_args: list[torch.SymInt, torch.Tensor]):
+        from thunder import jit
+
+        # Dynamo uses lazy generation of the underlying Python code, so we need to
+        # force recompilation of the GraphModule before passing it to Thunder.
+        gm.real_recompile()
+
+        # Here in the future we could add some logic to check if the GraphModule
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:57]: Great comment

File: thunder/dynamo/compiler.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        # will need to build a fallback mechanism to handle those cases.
+        # Possible stages of the compilation that need to be saved for inspection:
+        # 1. The GraphModule as it was passed to ThunderCompiler.
+        # 2. The GraphModule after split for Thunder/PyTorch.
+        # 3. If the whole GraphModule is not supported, record the reasons why.
+
+    def __call__(self, gm: torch.fx.GraphModule, sample_args: list[torch.SymInt, torch.Tensor]):
+        from thunder import jit
+
+        # Dynamo uses lazy generation of the underlying Python code, so we need to
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:53]: Nice comment

File: thunder/tests/framework.py [hunk 7]
```diff
@@ -232,10 +233,20 @@ def version(self):
         return torch.__version__
 
 
+class DynamoThunderTestExecutor(TestExecutor):
```
Comment (mruberry) [inline thunder/tests/framework.py:236]: This adds a test executor for Dynamo (which is cool), but doesn't register it, which means that tests generated to use test executors won't instantiate a variant for it.

That seems totally reasonable for now so we don't increase test time. Would you add a comment to this describing that it isn't returned from `_all_test_executors()` intentionally.

What are your thoughts on testing this variant? Did you have a chance to run the tests locally?

File: thunder/tests/test_dynamo.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
+import torch.fx
```
Comment (mruberry) [inline thunder/tests/test_dynamo.py:1]: Nice addition of a new test file

File: thunder/tests/test_networks.py [hunk 9]
```diff
@@ -10,16 +10,18 @@
...
-from thunder.tests.framework import instantiate, requiresCUDA
+from thunder.tests.framework import instantiate, requiresCUDA, DynamoThunderExecutor, _all_test_executors
 import thunder.tests.nanogpt_model as nanogpt_model
 import thunder.tests.hf_bart_self_attn as hf_bart_self_attn
 
 #
 # nanoGPT tests
 #
 
+all_test_executors_and_dynamo = _all_test_executors() + [DynamoThunderExecutor]
```
Comment (mruberry) [inline thunder/tests/test_networks.py:21]: Nice explicit addition. Would you add a comment for why it's needed?

File: thunder/dynamo/compiler.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        from thunder import ThunderModule
+
+        # Thunder-compiled functions should be readily available for inspection
+        # and testing, so we will store them in a list. The order of the
+        # functions in the list will be the same as the order in which they were
+        # compiled. In addition, we will store a mapping from the ThunderModule
+        # to the GraphModule that was passed to ThunderCompiler. This will allow
+        # us to inspect the GraphModule that was compiled by Thunder.
+        self.thunder_fns: list[ThunderModule] = []
+        self.thunder_to_gm: OrderedDict[ThunderModule, torch.fx.GraphModule] = OrderedDict()
```
Comment (crcrpar) [inline thunder/dynamo/compiler.py:39]: [question] iirc `dict` reserves the insertion order, so I'm wondering what/which features of `OrderedDict` are needed

File: thunder/dynamo/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from thunder.dynamo.compiler import ThunderCompiler
```
Comment (crcrpar) [inline thunder/dynamo/__init__.py:1]:
```suggestion
from thunder.dynamo.compiler import ThunderCompiler


__all__ = [
    "ThunderCompiler",
]
```

**Review by mruberry — APPROVED** [review]
Cool! A great step forward.

I made some minor suggestions for tweaks.

A cool follow-up would be a notebook demonstrating the backend for practitioners.

File: thunder/dynamo/compiler.py [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+        # There will be pieces of Dynamo IR that Thunder cannot compile, so we
+        # will need to build a fallback mechanism to handle those cases.
+        # Possible stages of the compilation that need to be saved for inspection:
+        # 1. The GraphModule as it was passed to ThunderCompiler.
+        # 2. The GraphModule after split for Thunder/PyTorch.
+        # 3. If the whole GraphModule is not supported, record the reasons why.
+
+    def __call__(self, gm: torch.fx.GraphModule, sample_args: list[torch.SymInt, torch.Tensor]):
+        from thunder import jit
```
Comment (mruberry) [inline thunder/dynamo/compiler.py:51]: Do we want to throw a warning (once) here that the backend is "experimental" or "in development"?

Comment (IvanYashchuk) [reply thunder/dynamo/compiler.py:51]: Yes, let's do that!

**Review by kshitij12345 — APPROVED** [review]
LGTM, thanks @IvanYashchuk

File: thunder/dynamo/compiler.py (hunk 12 above)
Comment (mruberry) [reply thunder/dynamo/compiler.py:51]: @apaz-cli has some functions to help warn once here:

https://github.com/Lightning-AI/lightning-thunder/blob/86cd9c3b5824064f8faa024ae0427fb297c04add/thunder/core/baseutils.py#L499

File: thunder/tests/framework.py (hunk 7 above)
Comment (IvanYashchuk) [reply thunder/tests/framework.py:236]: Thanks, I added a comment that this TestExecutor is not included in the `_all_test_executors` list and that test instantiation needs to opt-in to enable testing with this executor.

This variant is tested for some end-to-end tests we have in test_networks.py also it's indirectly tested through running nightly benchmarks (not part of the GitHub CI).

File: thunder/dynamo/compiler.py (hunk 12 above)
Comment (IvanYashchuk) [reply thunder/dynamo/compiler.py:55]: I added a warning that is shown each time when ThunderCompiler is initialized.

File: thunder/benchmarks/benchmark_litgpt.py [hunk 13]
```diff
@@ -393,7 +395,15 @@ def setup_compile(self, model):
 
                 executors.insert(0, transformer_engine_ex)
 
-            model = thunder.jit(model, executors=executors)
+            if "dynamo" in self.compile:
```
Comment (crcrpar) [inline thunder/benchmarks/benchmark_litgpt.py:398]: wouldn't we want to tweak the config like
https://github.com/Lightning-AI/lightning-thunder/blob/a206e22605e09f2817c1c5b33ff1fbfb28ada1b7/thunder/benchmarks/benchmark_litgpt.py#L367-L370?

Comment (IvanYashchuk) [reply thunder/benchmarks/benchmark_litgpt.py:398]: I don't know why this cache size limit was added.

Comment (crcrpar) [reply thunder/benchmarks/benchmark_litgpt.py:398]: hmm, in #940, torch.compile(fsdp2_model) seems to over the cache limit but the other configs look fine without the change.

**Review by crcrpar — APPROVED** [review]

**Comment by crcrpar** [comment]
How should I get thunder traces when I use thunder as torch.compile backend in this updated script?

**Comment by IvanYashchuk** [comment]
> How should I get thunder traces when I use thunder as torch.compile backend in this updated script?

First, instantiate our backend object `backend = ThunderCompiler(executors=executors)`, then after the first run of torch.compiled function Thunder functions will be available in `backend.thunder_fns`. `thunder.last_traces` can be called on each Thunder function from the `thunder_fns` list.

**Comment by t-vi** [comment]
So do we merge? From my POV it's ready to go. Thank you @IvanYashchuk @mruberry @crcrpar @kshitij12345

**Pull request merged**